File: Task/Hash-join/Haskell/hash-join-2.hs
{-# LANGUAGE TupleSections #-}
import qualified Data.Map as M
import Data.List
import Data.Maybe
import Control.Applicative

mapJoin xs fx ys fy = joined
  where yMap = foldl' f M.empty ys
        f m y = M.insertWith (++) (fy y) [y] m
        joined = concat . catMaybes .
                 map (\x -> map (x,) <$> M.lookup (fx x) yMap) $ xs

test = mapM_ print $ mapJoin
    [(1, "Jonah"), (2, "Alan"), (3, "Glory"), (4, "Popeye")]
        snd
    [("Jonah", "Whales"), ("Jonah", "Spiders"),
     ("Alan", "Ghosts"), ("Alan", "Zombies"), ("Glory", "Buffy")]
        fst

File: Task/Hash-join/Lua/hash-join-2.lua
local hashlist = {
  new = function(self,key)
    return setmetatable({key=key, hash={}, list={}}, {__index=self})
  end,
  insert = function(self,row)
    self.list[#self.list+1] = row
    if not self.hash[row[self.key]] then self.hash[row[self.key]]={} end
    table.insert(self.hash[row[self.key]], row)
    return self
  end,
  join = function(self,tabb)
    local result = {}
    for _,rowb in pairs(tabb.list) do
      if (self.hash[rowb[tabb.key]]) then
        for _,rowa in pairs(self.hash[rowb[tabb.key]]) do
          result[#result+1] = { A=rowa, B=rowb }
        end
      end
    end
    return result
  end
}

local function recA(age, name) return { Age=age, Name=name } end
tabA = hashlist:new("Name")
  :insert(recA(27,"Jonah"))
  :insert(recA(18,"Alan"))
  :insert(recA(28,"Glory"))
  :insert(recA(18,"Popeye"))
  :insert(recA(28,"Alan"))

local function recB(character, nemesis) return { Character=character, Nemesis=nemesis } end
local tabB = hashlist:new("Character")
  :insert(recB("Jonah","Whales"))
  :insert(recB("Jonah","Spiders"))
  :insert(recB("Alan","Ghosts"))
  :insert(recB("Alan","Zombies"))
  :insert(recB("Glory","Buffy"))

for _,row in pairs(tabA:join(tabB)) do
  print(row.A.Age, row.A.Name, row.B.Character, row.B.Nemesis)
end
print("or vice versa:")
for _,row in pairs(tabB:join(tabA)) do
  print(row.B.Age, row.B.Name, row.A.Character, row.A.Nemesis)
end
